client.go

```diff
@@ -925,12 +935,42 @@ func anyError(resp *http.Response, err error) error {
 }
 
 // doRequest creates and performs an http request.
-func (c *Client) doRequest(host *URI, method, path string, headers map[string]string, reader io.Reader) (*http.Response, error) {
-	req, err := makeRequest(host, method, path, headers, reader)
+func (c *Client) doRequest(host *URI, method, path string, headers map[string]string, data []byte) (*http.Response, error) {
+	var resp *http.Response
+	var err error
+	var req *http.Request
+	tries := 1 + c.retries
+	sleepTime := c.minRetrySleepTime
+	for tries > 0 {
+		req, err = makeRequest(host, method, path, headers, data)
+		if err != nil {
+			return nil, errors.Wrap(err, "building request")
+		}
+		tries--
+		resp, err = c.client.Do(req)
+		if err == nil {
+			if resp.StatusCode >= 200 && resp.StatusCode < 300 {
+				return resp, nil
+			}
+			// Pilosa nodes sometimes return 400, we retry in that case.
+			// No need to retry in other 4xx cases.
+			if resp.StatusCode > 400 && resp.StatusCode < 500 {
```

Comment:
I think we probably need to be more careful in Pilosa about what statuses we're returning, and which errors are retryable, and which are not. There are many cases with 400 errors where retrying doesn't make sense (unmarshal error at `pilosa/http/handler.go:1100`), and I think in some places we're returning 5XX errors, where it would make sense to retry (see `api.Import` at `pilosa/http/handler.go:1104`).

So... let's decide on one (or a set of) status codes that signify a temporary error condition where it's reasonable for the client to perform a retry on its own, and then let's make sure that Pilosa returns those status codes in the appropriate cases and does *not* return them when a retry doesn't make sense.

I think that the logic as it stands right now, won't actually retry if an import-roaring request fails due to the cluster being in state STARTING which is the motivating use case for this work.

Reply:
Actually it seems like it works quite nice when one of the nodes is in `STARTING` state. The instructions for reproducing https://github.com/pilosa/pilosa/issues/1585 in the #dev channel leaves the coordinator in `STARTING` state, and the client keeps retrying the import request. When the other node is restarted, the coordinator becomes `NORMAL` and the import progresses.

I agree with updating Pilosa to return better error codes BUT it's not a very big problem currently. Even if retrying doesn't make sense in some cases, it's also harmless.

We retry 500 errors.

Reply:
ah, I missed that we retry on 500... that makes more sense